RebbleOS/vibrate.c

```diff
@@ -48,70 +67,97 @@ void vibrate_init(void)
     _vibrate_queue = xQueueCreate(1, sizeof(uint8_t));
 }
 
-/*
+/**
+ * Execute a command
+ * @param command Command to execute. Note that VIBRATE_CMD_MAX is not a valid command.
+ */
+void vibrate_command(VibrateCmd_t command)
+{
+    if (command < VIBRATE_CMD_MAX)
+    {
+        vibrate_play_pattern(&_defaultVibratePatterns[command]);
+    }
+    else
+    {
+        // ya dun goofed
+        assert(false);
+    }
+}
+
+/**
+ * Play a given pattern.
+ * @param pattern Pointer to a struct containing a defined pattern.
+ */
+void vibrate_play_pattern(VibratePattern_t *pattern)
+{
+    pattern->curBufferIndex = 0;
+    (void) xQueueSendToBack(_vibrate_queue, &pattern, VIBRATE_QUEUE_MAX_WAIT_TICKS);   
+}
+
+/**
+ * Stop playing a pattern. If the _vibrate_thread is currently blocked by vTaskDelay,
+ * the pattern will be stopped when vTaskDelay returns.
+ */
+void vibrate_stop(void)
+{
+    
+    (void) xQueueSendToBack(_vibrate_queue, &_defaultVibratePatterns[VIBRATE_CMD_STOP], VIBRATE_QUEUE_MAX_WAIT_TICKS);    
+}
+
+/**
  * Start vibratin
  */
 static void _enable(uint8_t enabled)
 {
     hw_vibrate_enable(enabled);
 }
 
-/*
- * Quickly and dirtily pop out a pattern on the motor
+
+/**
+ * Play the currently selected pattern.
  */
-static void _play_pattern(uint16_t *pattern, uint8_t pattern_length)
+static void _play()
 {
-    // set the timer 12 CH2 to pwm to vibrate
-    // TODO FIX
-    return;
-    // low rumble
-    // touble tap
-    // medium
-    // high
-    // and drill
-    
-    for (uint8_t i = 0; i < pattern_length; i += 2)
-    {
-        // vibrate for x time
-        // until this is variable...
-        if (pattern[i] > 0)
-            _enable(1);
-        vTaskDelay(pattern[i+1] / portTICK_RATE_MS);
-        _enable(0);
+    while (currentPattern->curBufferIndex < currentPattern->length)
+    {        
+        if (currentPattern->buffer[currentPattern->curBufferIndex] > 0)
+        {
+            _enable(true);
+        }
+        
+        vTaskDelay(currentPattern->buffer[currentPattern->curBufferIndex] / portTICK_RATE_MS);
+        _enable(false);
+        
+        
+        if (xQueueReceive(_vibrate_queue, &currentPattern, 0))
+        {
+            // if we just received another pattern (including stop),
+            // start playing it from the beginning
+            currentPattern->curBufferIndex = 0;
+        }
+        else
+        {
+            currentPattern->curBufferIndex++;
+        }
     }
+    
+    currentPattern = &_defaultVibratePatterns[VIBRATE_CMD_STOP];
 }
 
 /*
  * The main task for driving patterns out of the motor
  */
 static void _vibrate_thread(void *pvParameters)
 {
-    uint8_t data;
-
+    currentPattern = &_defaultVibratePatterns[VIBRATE_CMD_STOP];
```

Comment:
I'll remove the _play function altogether, since it's it's really short and not being called elsewhere.

`current_pattern` will now be a local static variable 
(does it still need an underscore? `_current_pattern`)